lib/ZnapZend/ZFS.pm

```diff
@@ -1430,10 +1430,32 @@ sub fileExistsAndExec {
     my $remote;
 
     ($remote, $filePath) = $splitHostDataSet->($filePath);
-    my @ssh = $self->$buildRemote($remote, [@{$self->priv}, qw(test -x), $filePath]);
+    my @ssh1 = $self->$buildRemote($remote, [@{$self->priv}, qw(test -x), $filePath]);
+    # Note: with restricted shell setup, users may be unable to run
+    # fully qualified command names (with a slash) but may run what
+    # they have in the PATH.
+    my @ssh2 = $self->$buildRemote($remote, [@{$self->priv}, qw(command -v), $filePath]);
+
+    print STDERR '# ' . join(' ', @ssh1) . " || " . join(' ', @ssh2) . "\n" if $self->debug;
+    # Tricks to hide output of `command -v` from perl stdout:
+    if (!system(@ssh1)) {
+        return 1;
+    }
 
-    print STDERR '# ' . join(' ', @ssh) . "\n" if $self->debug;
-    return !system(@ssh);
+    my $ssh2out = `@ssh2`;
```

Comment:
hmmm this will invoke a shell locally, we try hard not todo that in the rest of the code ...

Reply:
Well, normally (in systems where existing code worked) it would not get to this line and `test-x` with `!system()` would be good enough, local or remote.

So this line and below only strike in situations where current code would assume `filePath` is absent (e.g. no `mbuffer` to run).

Is there a particular reason to avoid local shell?

Alternately, is there a good way to have similar effect (probably over SSH and with quirky setup of restricted shell) other than pulling in complex IPC3 or something of the sort?

Reply:
the problem is that shells have a tendency to be scrambled by user configuration and also to do unexpected things with the input ... the fewer shells the fewer problems ... we already got one at the other end of the ssh connection

Reply:
But here we rely exactly on POSIX (IIRC) built-in `command` and profile settings like `PATH`...

Reply:
lets try :)